### PR TITLE
feat: distributed execution of compact statement

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

distributed execution of compact statement

```sql
mysql> create table t_compact_0 (a int not null) row_per_block=5 block_per_segment=5;
Query OK, 0 rows affected (0.15 sec)

mysql> insert into t_compact_0 select 50 - number from numbers(100);
Query OK, 100 rows affected (0.42 sec)

mysql> insert into t_compact_0 select 50 - number from numbers(100);
Query OK, 100 rows affected (0.41 sec)

mysql> insert into t_compact_0 select 50 - number from numbers(100);
Query OK, 100 rows affected (0.31 sec)

mysql> select count(),sum(a) from t_compact_0;
+---------+--------+
| count() | sum(a) |
+---------+--------+
|     300 |    150 |
+---------+--------+
1 row in set (0.33 sec)
Read 300 rows, 1.17 KiB in 0.264 sec., 1.14 thousand rows/sec., 4.43 KiB/sec.

mysql> alter table t_compact_0 set options(row_per_block=10,block_per_segment=10);
Query OK, 0 rows affected (0.06 sec)

# lazy compact
# The number of compact segments task is greater than the number of cluster nodes, 
# so will build compact blocks task during pipeline init.
mysql> explain pipeline optimize table t_compact_0 compact;
+-------------------------------------------------------------------------------------------+
| explain                                                                                   |
+-------------------------------------------------------------------------------------------+
| CommitSink × 1 processor                                                                  |
|   TransformMergeCommitMeta × 1 processor                                                  |
|     TransformExchangeDeserializer × 1 processor                                           |
|       Merge (DummyTransform × 3 processors) to (TransformExchangeDeserializer × 1)        |
|         Merge (MutationAggregator × 1 processor) to (Resize × 3)                          |
|           MutationAggregator × 1 processor                                                |
|             Merge (TransformSerializeBlock × 20 processors) to (MutationAggregator × 1)   |
|               TransformSerializeBlock × 20 processors                                     |
|                 CompactSource × 20 processors                                             |
+-------------------------------------------------------------------------------------------+
9 rows in set (0.27 sec)
Read 0 rows, 0.00 B in 0.243 sec., 0 rows/sec., 0.00 B/sec.

mysql> optimize table t_compact_0 compact;
Query OK, 0 rows affected (0.31 sec)

mysql> select segment_count, block_count, row_count from fuse_snapshot('default', 't_compact_0') limit 2;
+---------------+-------------+-----------+
| segment_count | block_count | row_count |
+---------------+-------------+-----------+
|             3 |          30 |       300 |
|             6 |          30 |       300 |
+---------------+-------------+-----------+
2 rows in set (0.09 sec)
Read 2 rows, 428.00 B in 0.055 sec., 36.58 rows/sec., 7.64 KiB/sec.
```

```sql
mysql> create table t_compact_1 (a int not null) row_per_block=5 block_per_segment=5;
Query OK, 0 rows affected (0.14 sec)

mysql> insert into t_compact_1 select 100 - number from numbers(150);
Query OK, 150 rows affected (0.42 sec)

mysql> alter table t_compact_1 set options(row_per_block=10,block_per_segment=15);
Query OK, 0 rows affected (0.11 sec)

# nolazy compact
# The number of compact segments task is less than the number of cluster nodes, 
# so will build compact blocks task before execute pipeline
mysql> explain pipeline optimize table t_compact_1 compact;
+----------------------------------------------------------------------------------------+
| explain                                                                                |
+----------------------------------------------------------------------------------------+
| CommitSink × 1 processor                                                               |
|   MutationAggregator × 1 processor                                                     |
|     Merge (TransformExchangeDeserializer × 6 processors) to (MutationAggregator × 1)   |
|       TransformExchangeDeserializer × 6 processors                                     |
|         Merge (DummyTransform × 8 processors) to (TransformExchangeDeserializer × 6)   |
|           Merge (TransformSerializeBlock × 6 processors) to (Resize × 8)               |
|             TransformSerializeBlock × 6 processors                                     |
|               CompactSource × 6 processors                                             |
+----------------------------------------------------------------------------------------+
8 rows in set (0.17 sec)
Read 0 rows, 0.00 B in 0.159 sec., 0 rows/sec., 0.00 B/sec.

mysql> optimize table t_compact_1 compact;
Query OK, 150 rows affected (0.34 sec)

mysql> select segment_count, block_count, row_count from fuse_snapshot('default', 't_compact_1') limit 2;
+---------------+-------------+-----------+
| segment_count | block_count | row_count |
+---------------+-------------+-----------+
|             2 |          14 |       150 |
|             6 |          30 |       150 |
+---------------+-------------+-----------+
2 rows in set (0.09 sec)
Read 2 rows, 396.00 B in 0.055 sec., 36.15 rows/sec., 6.99 KiB/sec.
```

- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12750)

### CODE REVIEW

**Comment by zhyass** [comment]
askbend:summary

File: tests/sqllogictests/suites/mode/cluster/distributed_compact.sql [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+select count(),sum(a) from t_compact
+----
+200 100
+
+statement ok
+alter table t_compact set options(row_per_block=10,block_per_segment=10)
+
+# lazy compact
+statement ok
+optimize table t_compact compact
```
Comment (BohuTANG) [inline tests/sqllogictests/suites/mode/cluster/distributed_compact.sql:23]: How to know the compact is in distributed or not(For the query we can explain)?

**Comment by SkyFan2002** [comment]
LGTM

**Review by SkyFan2002 — APPROVED** [review]

File: src/query/storages/fuse/src/operations/mutation/compact/compact_part.rs [hunk 2]
```diff
@@ -13,23 +13,67 @@
...
+}
 
 #[derive(serde::Serialize, serde::Deserialize, PartialEq)]
-pub struct CompactPartInfo {
-    pub blocks: Vec<Arc<BlockMeta>>,
-    pub index: BlockMetaIndex,
+pub enum CompactPartInfo {
+    CompactExtraInfo(CompactExtraInfo),
+    CompactTaskInfo(CompactTaskInfo),
 }
```
Comment (JackTan25) [inline src/query/storages/fuse/src/operations/mutation/compact/compact_part.rs:74]: what's the difference of CompactExtraInfo and CompactTaskInfo?

File: src/query/storages/fuse/src/operations/mutation/compact/compact_part.rs [hunk 3]
```diff
@@ -13,23 +13,67 @@
...
 
 use crate::operations::common::BlockMetaIndex;
+use crate::operations::mutation::BlockIndex;
+use crate::operations::mutation::SegmentIndex;
+
+#[derive(serde::Serialize, serde::Deserialize, PartialEq, Clone)]
+pub struct CompactLazyPartInfo {
+    pub segment_indices: Vec<SegmentIndex>,
+    pub compact_segments: Vec<Arc<CompactSegmentInfo>>,
+}
```
Comment (JackTan25) [inline src/query/storages/fuse/src/operations/mutation/compact/compact_part.rs:37]: let me see, CompactLazyPartInfo is used to see if this execution is distributed, right?  sorry, I'm a little confused, what's the difference of CompactLazyPartInfo and CompactPartInfo?

Comment (zhyass) [reply src/query/storages/fuse/src/operations/mutation/compact/compact_part.rs:37]: Just like `FuseLazyPartInfo` and `FusePartInfo`.

File: src/query/storages/fuse/src/operations/mutation/compact/compact_part.rs (hunk 2 above)
Comment (zhyass) [reply src/query/storages/fuse/src/operations/mutation/compact/compact_part.rs:74]: > what's the difference of CompactExtraInfo and CompactTaskInfo?

CompactExtraInfo: the extra info for compact, such as unchanged blocks, removed segment indices, statistics.
CompactTaskInfo: the compact task

**Review by dantengsky — APPROVED** [review]

**Pull request merged**